### DISCUSSION TITLE
Borg and borg_think_dungeon_brave

### DISCUSSION THREAD

**Discussion by Netbrian** [angband/angband#6096] [General]
I've managed to get Angband to compile with Visual Studio!

Encouraging me to do this may have been a mistake on everyone's part, because now I have more questions.

I've been running the debugger and trying to wrap my head around the code.  I noticed that the borg was calling borg_think_dungeon_brave a lot.  It's says it's used when the borg is "stuck" somehow.

However, in many cases the dungeon seems perfectly navigable?  In this example, there are plenty of places he could go without being "brave."

[PLAYER.zip](https://github.com/user-attachments/files/17356724/PLAYER.zip)

Thank you!

@agoodman00

**Comment by agoodman00** [comment]
> I've managed to get Angband to compile with Visual Studio!

Most Excellent!  I am always glad when more people are playing with the borg code. 
 
> Encouraging me to do this may have been a mistake on everyone's part, because now I have more questions.

Nothing wrong with questions as long as you don't mind me taking a bit to answer :) 
 
> I've been running the debugger and trying to wrap my head around the code. I noticed that the borg was calling borg_think_dungeon_brave a lot. It's says it's used when the borg is "stuck" somehow.
> 
> However, in many cases the dungeon seems perfectly navigable? In this example, there are plenty of places he could go without being "brave."

I can take a look at this case but generally there are two most common reasons the borg can run out of things to do when there are obviously things to do... 

1) the borg is scared of an area.  He got hit with some attack there and he marked the area as "scary" and doesn't want to go there.  If he would have to walk through a "scary" area to get to a thing to do (explore unexplored area, attack someone, pickup stuff etc) he needs to pump up his bravery before he will do it... risk death rather than just sitting around.  This is talked about in the comment above borg_caution() in borg_caution.c and is tracked primarily in the borg_fear_region and borg_fear_monsters arrays.

2) the borg was retrofitted with a "leash" at some point.  This leash measures the distance between the borg and the stairs and doesn't let him go too far (distance determined by clevel, IIRC... I would have to look at the code to remember for sure) from the stairs so he can easily run away when things get dangerous.  He often looks like he has things to do but if they are beyond leash distance, he can't do them and eventually just gives up and heads up the stairs.
I am not in love with this processing and want to add some code to make sure he is better at exploring his area when on a leash since he seems to repeat small explorations over and over but there isn't any doubt that it reduces early deaths.

**Comment by agoodman00** [comment]
Ya, this looks like 2 above.  
Mostly I think when you are saying "the dungeon seems perfectly navigable" you mean "there are unexplored areas to explore".  You can take a look at borg_flow_dark() in borg-flow-dark.c and this is where he decides if the dark (unexplored) area is a place he can explore.  In the depths of that code you can see things like 
    /* check the leash length */
    if (borg.trait[BI_CDEPTH] >= borg.trait[BI_CLEVEL] - 5)
        leash = borg.trait[BI_CLEVEL] * 3 + 9;
(make the leash 3* charater level + 9 if you are below char level - 5 ... in other words "don't go too far from the stairs if you are deeper than is very safe")
and 
            /* don't go too far from the stairs */
            if (borg_flow_far_from_stairs_dist(x, y, b_stair, leash))
                continue;
(check how far this is from the stairs, leash dependent)

**Reply by Netbrian** [reply]
OK, this makes sense.

I'll keep an eye on this more -- at high bravely levels, the borg seems to behave oddly, so we might want to figure out another way to handle the leash.

Thank you!

**Reply by agoodman00** [reply]
> OK, this makes sense.
> 
> 
> 
> I'll keep an eye on this more -- at high bravely levels, the borg seems to behave oddly, so we might want to figure out another way to handle the leash.
> 
> 
> 
> Thank you!

I agree that there must be better ways to handle keeping low level Borgs alive. I have considered having a specialized borg_think_dungeon_low_level() that handles that but it would be a lot of duplicate code. So far it hasn't bubbled up to the most important borg code to play with

**Comment by Netbrian** [comment]
Which part of the code helps the borg set goals on where to do next?

**Reply by agoodman00** [reply]
> Which part of the code helps the borg set goals on where to do next?

In borg_think_dungeon() there are a bunch of borg_flow_xxx (borg_flow_dark(), borg_flow_kill() etc). During those the goal is set and one step toward the goal is taken for walking goals. There are also routines for shopping goals or resting or a few other things.

**Comment by Netbrian** [comment]
Is there a way to set a breakpoint when the borg encounters a particular unique?

Right now, I have a priest that has been stuck on Eol for over a day.  I'd like to see why he might be struggling with that particular encounter, but can't figure out a way to see what's going on.

**Comment by agoodman00** [comment]
breakpoint... hmmm, not sure but I think you can add a "search" for it though I haven't tested the functionality of that code in 20 years.  That is ^zs then type in a "match string" and if that string comes up in any message the borg SHOULD stop.
now for breakpoints... you would need to break at about borg_flow_kill.c line 2827 and know the race index (r_idx) of that unique or break on r_ptr->name[0] == 'E' && r_ptr->name[2] == 'l' && r_ptr->name[3] == ',' 
(note: left off the ö because I don't know how happy the debugger would be with accents)
